### PR TITLE
 Implement SSE2 StoreNonTemporal HW intrinsic - complete SSE2 ISA

### PR DESCRIPTION
**Pull request by 4creators**
This PR is independent of #16797 and #16817, however, it is assumed to complete SSE2 HW intrinsics and sets SSE2 to be complete ISA and should be committed as a last one.

FYI @CarolEidt @fiigii @tannergooding

### CODE REVIEW

File: src/jit/hwintrinsicxarch.cpp [hunk 1]
```diff
@@ -1030,6 +1030,23 @@ GenTree* Compiler::impSSE2Intrinsic(NamedIntrinsic        intrinsic,
...
+        {
+            assert(sig->numArgs == 2);
+            assert(JITtype2varType(sig->retType) == TYP_VOID);
+            op2      = impPopStack().val;
+            op1      = impPopStack().val;
+            simdSize = ((op2->TypeGet() == TYP_LONG) || (op2->TypeGet() == TYP_ULONG)) ? 8 : 4;
+#ifdef _TARGET_X86_
+            if (simdSize == 8)
+            {
+                return impUnsupportedHWIntrinsic(CORINFO_HELP_THROW_PLATFORM_NOT_SUPPORTED, method, sig, mustExpand);
```
Comment (fiigii) [inline src/jit/hwintrinsicxarch.cpp]: You do not need this code. Just give this intrinsic a flag `HW_Flag_SecondArgMaybe64Bit` is enough.

File: src/jit/hwintrinsicxarch.cpp [hunk 2]
```diff
@@ -1030,6 +1030,23 @@ GenTree* Compiler::impSSE2Intrinsic(NamedIntrinsic        intrinsic,
             break;
         }
 
+        case NI_SSE2_StoreNonTemporal:
+        {
+            assert(sig->numArgs == 2);
+            assert(JITtype2varType(sig->retType) == TYP_VOID);
+            op2      = impPopStack().val;
+            op1      = impPopStack().val;
+            simdSize = ((op2->TypeGet() == TYP_LONG) || (op2->TypeGet() == TYP_ULONG)) ? 8 : 4;
```
Comment (fiigii) [inline src/jit/hwintrinsicxarch.cpp]: `simdSize` should be `0` for scalar intrinsics.

File: src/jit/hwintrinsicxarch.cpp [hunk 3]
```diff
@@ -1030,6 +1030,23 @@ GenTree* Compiler::impSSE2Intrinsic(NamedIntrinsic        intrinsic,
...
+#ifdef _TARGET_X86_
+            if (simdSize == 8)
+            {
+                return impUnsupportedHWIntrinsic(CORINFO_HELP_THROW_PLATFORM_NOT_SUPPORTED, method, sig, mustExpand);
+            }
+#else
+            retNode  = gtNewSimdHWIntrinsicNode(TYP_VOID, op1, op2, NI_SSE2_StoreNonTemporal, op2->TypeGet(), simdSize);
```
Comment (fiigii) [inline src/jit/hwintrinsicxarch.cpp]: Please use `gtNewScalarHWIntrinsicNode` constructor.

File: src/jit/hwintrinsiclistxarch.h [hunk 4]
```diff
@@ -230,6 +230,7 @@ HARDWARE_INTRINSIC(SSE2_StoreAligned,                                "StoreAlign
...
 HARDWARE_INTRINSIC(SSE2_StoreLow,                                    "StoreLow",                                         SSE2,       -1,           16,          2,            {INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_movq,      INS_movq,      INS_invalid,   INS_movlpd},            HW_Category_MemoryStore,                       HW_Flag_NoRMWSemantics)
+HARDWARE_INTRINSIC(SSE2_StoreNonTemporal,                            "StoreNonTemporal",                                 SSE2,       -1,           16,          2,            {INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_movnti,    INS_movnti,    INS_movnti,    INS_movnti,    INS_invalid,   INS_invalid},           HW_Category_Special,                           HW_Flag_NoRMWSemantics)
```
Comment (fiigii) [inline src/jit/hwintrinsiclistxarch.h]: This should have the category of `HW_Category_Scalar`.

**Comment by 4creators** [comment]
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

**Comment by 4creators** [comment]
test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

**Comment by 4creators** [comment]
@CarolEidt @fiigii @tannergooding

PR is ready for review PTAL

File: src/jit/hwintrinsiccodegenxarch.cpp [hunk 5]
```diff
@@ -1031,6 +1031,18 @@ void CodeGen::genSSE2Intrinsic(GenTreeHWIntrinsic* node)
             break;
         }
 
+        case NI_SSE2_StoreNonTemporal:
```
Comment (tannergooding) [inline src/jit/hwintrinsiccodegenxarch.cpp]: Why can't this be handled by the table driven 2-op MemoryStore case?

Comment (4creators) [reply src/jit/hwintrinsiccodegenxarch.cpp]: Essentially it is a scalar intrinsic and @fiigii asked specifically to put it in this category. Currently scalar intrinsics are not table driven.

Comment (4creators) [reply src/jit/hwintrinsiccodegenxarch.cpp]: https://github.com/dotnet/coreclr/pull/16832#discussion_r173096844

Comment (tannergooding) [reply src/jit/hwintrinsiccodegenxarch.cpp]: Ah, I see. We can fix this later.
Thanks

Comment (fiigii) [reply src/jit/hwintrinsiccodegenxarch.cpp]: > . We can fix this later.

Implementing `Bmi1/2` is a good opportunity.

File: tests/src/JIT/HardwareIntrinsics/X86/Sse2/StoreNonTemporal.cs [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        const int Pass = 100;
+        const int Fail = 0;
+
+        static unsafe int Main(string[] args)
+        {
+            int testResult = Pass;
+
+            if (Sse2.IsSupported)
+            {
+                if (Environment.Is64BitProcess)
```
Comment (fiigii) [inline tests/src/JIT/HardwareIntrinsics/X86/Sse2/StoreNonTemporal.cs]: It is better to add a scenario that tests the exception on 32-bit platforms.

File: tests/src/JIT/HardwareIntrinsics/X86/Sse2/StoreNonTemporal.cs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+                    {
+                        long* inArray = stackalloc long[2];
+                        inArray[0] = 0xffffffff01l;
+                        long* outBuffer = stackalloc long[2];
+
+                        Sse2.StoreNonTemporal(outBuffer, inArray[0]);
+                        testResult = Fail;
+                        Console.WriteLine($"{nameof(Sse2)}.{nameof(Sse2.StoreNonTemporal)} failed on long: expected PlatformNotSupportedException exception.");
+                    }
+                    catch (PlatformNotSupportedException)
```
Comment (fiigii) [inline tests/src/JIT/HardwareIntrinsics/X86/Sse2/StoreNonTemporal.cs]: 👍

**Comment by tannergooding** [comment]
@4creators, could you rebase and retest?

**Comment by 4creators** [comment]
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

**Comment by 4creators** [comment]
test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

**Comment by 4creators** [comment]
All Ubuntu builds timed out.

**Comment by tannergooding** [comment]
Just to explicitly call it out, this should not be merged until after the following two PRs are also merged:
* https://github.com/dotnet/coreclr/pull/16817
* https://github.com/dotnet/coreclr/pull/16797

**Comment by 4creators** [comment]
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

**Comment by 4creators** [comment]
test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2

**Comment by 4creators** [comment]
@CarolEidt PTAL

Test failure is due to known issue, see above.

**Comment by 4creators** [comment]
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2

**Comment by 4creators** [comment]
This time Ubuntu build timed out.

**Comment by 4creators** [comment]
```
03:32:15.600     Finished:    JIT.SIMD.XUnitWrapper
03:32:17.144     Finished:    JIT.Performance.XUnitWrapper
03:33:15.859     Finished:    JIT.opt.XUnitWrapper
03:33:17.609     Finished:    JIT.CodeGenBringUpTests.XUnitWrapper
03:35:32.593     Finished:    JIT.Generics.XUnitWrapper
03:38:19.498     Finished:    JIT.HardwareIntrinsics.XUnitWrapper
03:38:20.611     Finished:    GC.Scenarios.XUnitWrapper
03:38:40.347     Finished:    JIT.IL_Conformance.XUnitWrapper
03:45:30.374     Finished:    JIT.Directed.XUnitWrapper
03:50:04.508     Finished:    JIT.jit64.XUnitWrapper
03:53:49.806     Finished:    baseservices.threading.XUnitWrapper
03:54:09.882     Finished:    JIT.Regression.XUnitWrapper
04:00:34.032 
Build timed out (after 240 minutes). Marking the build as aborted.
04:00:34.166 
Build was aborted
```

**Comment by 4creators** [comment]
@CarolEidt It is possible to merge this PR as well as all PRs on which it depends are in. Could you pls take a look?

**Comment by CarolEidt** [comment]
test Ubuntu x64 Checked jitincompletehwintrinsic

**Comment by CarolEidt** [comment]
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2

**Pull request merged**